README.md
JS-YAML - YAML 1.1 parser for JavaScript
========================================

[![Build Status](https://secure.travis-ci.org/nodeca/js-yaml.png)](http://travis-ci.org/nodeca/js-yaml)

[Online Demo](http://nodeca.github.com/js-yaml/)

This is a native port of [PyYAML](http://pyyaml.org/), the most advanced YAML parser.
Now you can use all modern YAML feature right in JavaScript. Originally snapshoted version - PyYAML 3.10 (2011-05-30).

## Installation

### YAML module for node.js

```
npm install js-yaml
```

If you want to inspect your YAML files from CLI, install [js-yaml.bin](https://github.com/nodeca/js-yaml.bin)
and then you'll be able to call it like this:

```
$ jsyaml /home/ixti/example.yml 
---------------------------------------------------------------------------
{ jsyaml: 'rocks!' }
...........................................................................
```


### bundled YAML library for browser

``` html
<script src="js-yaml.min.js"></script>
<script type="text/javascript">
var doc = jsyaml.load('greeting: hello\nname: world');
</script>
```

Also we support AMD loaders, e.g. [RequireJS](http://requirejs.org/).

## API

JS-YAML automatically registers handlers for `.yml` and `.yaml` files. You can load them just with `require`.
That's mostly equivalent to calling loadAll() on file handler and picking out the first found document.
Notice, if the file has more than one document, this will silently ignore all
the documents except the first one.

``` javascript
require('js-yaml');

// Get array of documents, or throw exception on error
var doc = require('/home/ixti/example.yml');

console.log(doc);
```


### load (string|buffer|file\_resource)

Parses source as single YAML document. Returns JS object or throws exception on error.

This function does NOT understands multi-doc sources, it throws exception on those.

``` javascript
var yaml = require('js-yaml');

// pass the string
fs.readFile('/home/ixti/example.yml', 'utf8', function (err, data) {
  if (err) {
    // handle error
    return;
  }
  try {
    console.log( yaml.load(data) );
  } catch(e) {
    console.log(e);
  }
});
```


### loadAll (string|buffer|file\_resource, iterator)

Same as `Load`, but understands multi-doc sources and apply iterator to each document.

``` javascript
var yaml = require('js-yaml');

// pass the string
fs.readFile('/home/ixti/example.yml', 'utf8', function (err, data) {
  if (err) {
    // handle error
    return;
  }

  try {
    yaml.loadAll(data, function (doc) {
      console.log(doc);
    });
  } catch(e) {
    console.log(e);
  }
});
```


### safeLoad (string|buffer|file\_resource)

Same as `load()` but uses _safe_ schema - only recommended tags of YAML
specification (no JavaScript-specific tags, e.g. `!!js/regexp`).


### safeLoadAll (string|buffer|file\_resource, iterator)

Same as `loadAll()` but uses _safe_ schema - only recommended tags of YAML
specification (no JavaScript-specific tags, e.g. `!!js/regexp`).


## JavaScript YAML tags scheme

The list of standard YAML tags and corresponding JavaScipt types. See also
[YAML Tag Discussion](http://pyyaml.org/wiki/YAMLTagDiscussion) and [Yaml Types](http://yaml.org/type/).

```
!!null ''                   # null
!!bool 'yes'                # bool
!!int '3...'                # number
!!float '3.14...'           # number
!!binary '...base64...'     # buffer
!!timestamp 'YYYY-...'      # date
!!omap [ ... ]              # array of key-value pairs
!!pairs [ ... ]             # array or array pairs
!!set { ... }               # array of objects with given keys and null values
!!str '...'                 # string
!!seq [ ... ]               # array
!!map { ... }               # object
```

**JavaScript-specific tags**

```
!!js/regexp /pattern/gim            # RegExp
!!js/undefined ''                   # Undefined
!!js/function 'function () {...}'   # Function
```

### Caveats

Note, that if you use arrays as key in JS, it's automatically converted to string.
So, if you have maps as key in YAML, result will flat:

``` yaml
---
? - foo
  - bar
: - baz
```

=>

``` javascript
{ "foo,bar": ["baz"] }
```

## License

View the [LICENSE](https://github.com/nodeca/js-yaml/blob/master/LICENSE) file (MIT).


HISTORY.md
1.0.0 / ????-??-??
------------------

* require now returns first found document in the stream instead of array
  of documents. See #10.

0.3.4 / ????-??-??
------------------

* improves compact output of Error. Closes #33.

0.3.3 / 2011-12-20
------------------

* jsyaml executable moved to separate module.
* adds `compact` stringification of Errors.

0.3.2 / 2011-12-16
------------------

* Fixes ug with block style scalars. Closes #26.
* All sources are passing JSLint now.
* Fixes bug in Safari. Closes #28.
* Fixes bug in Opers. Closes #29.
* Improves browser support. Closes #20.
* Added jsyaml executable.
* Added !!js/function support. Closes #12.

0.3.1 / 2011-11-18
------------------

* Added AMD support for browserified version.
* Wrapped browserified js-yaml into closure.
* Fixed the resolvement of non-specific tags. Closes #17.
* Added permalinks for online demo YAML snippets. Now we have YPaste service, lol.
* Added !!js/regexp and !!js/undefined types. Partially solves #12.
* Fixed !!set mapping.
* Fixed month parse in dates. Closes #19.

0.3.0 / 2011-11-09
------------------

* Removed JS.Class dependency. Closes #3.
* Added browserified version. Closes #13.
* Added live demo of browserified version.
* Ported some of the PyYAML tests. See #14.
* Fixed timestamp bug when fraction was given.

0.2.2 / 2011-11-06
------------------

* Fixed crash on docs without ---. Closes #8.
* Fixed miltiline string parse
* Fixed tests/comments for using array as key

0.2.1 / 2011-11-02
------------------

* Fixed short file read (<4k). Closes #9.

0.2.0 / 2011-11-02
------------------

* First public release


test/issues/issue-8.js
var Assert = require('assert');
var JsYaml = require('../../lib/js-yaml');
var source = __dirname + '/data/issue-8.yml';

module.exports = require('../helper').issue({
  title: "#8: Parse failed when no document start present",
  fixed: true,
  test: function () {
    Assert.doesNotThrow(function () {
      require(source);
    }, TypeError);
  }
});


test/issues/issue-17.js
var Assert = require('assert');
var JsYaml = require('../../lib/js-yaml');
var source = __dirname + '/data/issue-17.yml';


module.exports = require('../helper').issue({
  title: "#17: Non-specific `!` tags should resolve to !!str",
  fixed: true,
  test: function () {
    var str = require(source);
    Assert.equal('string', typeof str);
  }
});


test/issues/issue-19.js
var Assert = require('assert');
var JsYaml = require('../../lib/js-yaml');
var source = __dirname + '/data/issue-19.yml';


module.exports = require('../helper').issue({
  title: "#19: Timestamp parsing is one month off",
  fixed: true,
  test: function () {
    var doc = require(source), expected = new Date(2011, 11, 24);

    // JS month starts with 0 (0 => Jan, 1 => Feb, ...)
    Assert.equal(doc.xmas.getTime(), expected.getTime());
  }
});


lib/js-yaml.js
'use strict';


var fs = require('fs');
var _loader = require('./js-yaml/loader');


var jsyaml = module.exports = {};


jsyaml.scan = function scan(stream, callback, Loader) {
  Loader = (Loader || _loader.SafeLoader);

  var loader = new Loader(stream);
  while (loader.checkToken()) {
    callback(loader.getToken());
  }
};


jsyaml.compose = function compose(stream, Loader) {
  Loader = (Loader || _loader.SafeLoader);

  var loader = new Loader(stream);
  return loader.getSingleNode();
};


jsyaml.load = function load(stream, Loader) {
  Loader = (Loader || _loader.Loader);

  var loader = new Loader(stream);
  return loader.getSingleData();
};


jsyaml.loadAll = function loadAll(stream, callback, Loader) {
  Loader = (Loader || _loader.Loader);

  var loader = new Loader(stream);
  while (loader.checkData()) {
    callback(loader.getData());
  }
};


jsyaml.safeLoad = function load(stream) {
  return jsyaml.load(stream, _loader.SafeLoader);
};


jsyaml.safeLoadAll = function loadAll(stream, callback) {
  jsyaml.loadAll(stream, callback, _loader.SafeLoader);
};


/**
 *  jsyaml.addConstructor(tag, constructor[, Loader]) -> Void
 *  Add a constructor for the given tag.
 *
 *  Constructor is a function that accepts a Loader instance
 *  and a node object and produces the corresponding JavaScript object.
 **/
jsyaml.addConstructor = function addConstructor(tag, constructor, Loader) {
  (Loader || _loader.Loader).addConstructor(tag, constructor);
};


// Register extensions handler
(function () {
  var require_handler = function (module, filename) {
    var fd = fs.openSync(filename, 'r'), docs = [];
    
    // fill in documents
    jsyaml.loadAll(fd, function (doc) { docs.push(doc); });
    module.exports = docs.shift();

    fs.closeSync(fd);
  };

  // register require extensions only if we're on node.js
  // hack for browserify
  if (undefined !== require.extensions) {
    require.extensions['.yml'] = require_handler;
    require.extensions['.yaml'] = require_handler;
  }
}());


////////////////////////////////////////////////////////////////////////////////
// vim:ts=2:sw=2
////////////////////////////////////////////////////////////////////////////////


lib/js-yaml/dumper.js
'use strict';


var $$ = require('./core'),
    _emitter = require('./emitter'),
    _serializer = require('./serializer'),
    _representer = require('./representer'),
    _resolver = require('./resolver');


function BaseDumper(stream, defaultStyle, defaultFlowStyle, canonical, indent, width,
                    allowUnicode, lineBreak, encoding, explicitStart, explicitEnd, version, tags) {
  _emitter.Emitter.call(this, stream, canonical, indent, width, allowUnicode, lineBreak);
  _serializer.Serializer.call(this, encoding, explicitStart, explicitEnd, version, tags);
  _representer.BaseRepresenter.call(this, defaultStyle, defaultFlowStyle);
  _resolver.BaseResolver.call(this);
}

$$.extend(BaseDumper.prototype,
          _emitter.Emitter.prototype,
          _serializer.Serializer.prototype,
          _representer.BaseRepresenter.prototype,
          _resolver.BaseResolver.prototype);


function SafeDumper(stream, defaultStyle, defaultFlowStyle, canonical, indent, width,
                    allowUnicode, lineBreak, encoding, explicitStart, explicitEnd, version, tags) {
  _emitter.Emitter.call(this, stream, canonical, indent, width, allowUnicode, lineBreak);
  _serializer.Serializer.call(this, encoding, explicitStart, explicitEnd, version, tags);
  _representer.SafeRepresenter.call(this, defaultStyle, defaultFlowStyle);
  _resolver.SafeResolver.call(this);
}

$$.extend(SafeDumper.prototype,
          _emitter.Emitter.prototype,
          _serializer.Serializer.prototype,
          _representer.SafeRepresenter.prototype,
          _resolver.SafeResolver.prototype);


function Dumper(stream, defaultStyle, defaultFlowStyle, canonical, indent, width,
                allowUnicode, lineBreak, encoding, explicitStart, explicitEnd, version, tags) {
  _emitter.Emitter.call(this, stream, canonical, indent, width, allowUnicode, lineBreak);
  _serializer.Serializer.call(this, encoding, explicitStart, explicitEnd, version, tags);
  _representer.Representer.call(this, defaultStyle, defaultFlowStyle);
  _resolver.Resolver.call(this);
}

$$.extend(Dumper.prototype,
          _emitter.Emitter.prototype,
          _serializer.Serializer.prototype,
          _representer.Representer.prototype,
          _resolver.Resolver.prototype);


module.exports.BaseDumper = BaseDumper;
module.exports.SafeDumper = SafeDumper;
module.exports.Dumper = Dumper;


////////////////////////////////////////////////////////////////////////////////
// vim:ts=2:sw=2
////////////////////////////////////////////////////////////////////////////////


lib/js-yaml/serializer.js
'use strict';


var $$ = require('./core'),
    _errors = require('./errors'),
    _events = require('./events'),
    _nodes = require('./nodes');


function SerializerError() {
  _errors.YAMLError.apply(this, arguments);
  this.name = 'SerializerError';
}
$$.inherits(SerializerError, _errors.YAMLError);


function Serializer(encoding, explicitStart, explicitEnd, version, tags) {
  this.useEncoding = encoding || null;
  this.useExplicitStart = explicitStart || null;
  this.useExplicitEnd = explicitEnd || null;
  this.useVersion = version || null;
  this.useTags = tags || null;
  this.serializedNodes = new $$.Hash();
  this.anchors = new $$.Hash();
  this.lastAnchorId = 0;
  this.closed = null;
}


Serializer.prototype.open = function open() {
  if (null === this.closed) {
    this.emit(new _events.StreamStartEvent(this.useEncoding));
    this.closed = false;
  } else if (this.closed) {
    throw new SerializerError("Serializer is closed");
  } else {
    throw new SerializerError("Serializer is already opened");
  }
};


Serializer.prototype.close = function close() {
  if (null === this.closed) {
    throw new SerializerError("Serializer is not opened");
  } else if (!this.closed) {
    this.emit(new _events.StreamEndEvent());
    this.closed = true;
  }
};


Serializer.prototype.serialize = function serialize(node) {
  if (null === this.closed) {
    throw new SerializerError("Serializer is not opened");
  } else if (this.closed) {
    throw new SerializerError("Serializer is closed");
  }

  this.emit(new _events.DocumentStartEvent(this.useExplicitStart, this.version, this.useTags));
  this.anchorNode(node);
  this.serializeNode(node);
  this.emit(new _events.DocumentEndEvent(this.useExplicitEnd));

  this.serializedNodes = new $$.Hash();
  this.anchors = new $$.Hash();
  this.lastAnchorId = 0;
};


Serializer.prototype.anchorNode = function anchorNode(node) {
  if (this.anchors.hasKey(node)) {
    if (null === this.anchors.get(node)) {
      this.anchors.store(node, this.generateAnchor(node));
    }
  } else {
    this.anchors.store(node, null);

    if ($$.isInstanceOf(node, _nodes.SequenceNode)) {
      $$.each(node.value, function (subnode) {
        this.anchorNode(subnode);
      }.bind(this));
    } else if ($$.isInstanceOf(node, _nodes.MappingNode)) {
      $$.each(node.value, function (pairs) {
        this.anchorNode(pairs[0]);
        this.anchorNode(pairs[1]);
      }.bind(this));
    }
  }
};


Serializer.prototype.generateAnchor = function generateAnchor(node) {
  this.lastAnchorId += 1;
  return 'id000'.slice(0, -this.lastAnchorId.toString().length) + this.lastAnchorId;
};


Serializer.prototype.serializeNode = function serializeNode(node, parent, index) {
  var alias, detectedTag, defaultTag, implicit;

  alias = this.anchors.get(node);

  if (this.serializedNodes.hasKey(node)) {
    this.emit(new _events.AliasEvent(alias));
    return;
  }

  this.serializedNodes.store(node, true);

  if ($$.isInstanceOf(node, _nodes.ScalarNode)) {
    detectedTag = this.resolve(_nodes.ScalarNode, node.value, [true, false]);
    defaultTag = this.resolve(_nodes.ScalarNode, node.value, [false, true]);
    implicit = [(node.tag === detectedTag), (node.tag === defaultTag)];
    this.emit(new _events.ScalarEvent(alias, node.tag, implicit, node.value, node.style));
  } else if ($$.isInstanceOf(node, _nodes.SequenceNode)) {
      implicit = (node.tag === this.resolve(_nodes.SequenceNode, node.value, true));
      this.emit(new _events.SequenceStartEvent(alias, node.tag, implicit, node.flowStyle));

      index = 0;
      $$.each(node.value, function (subnode) {
        this.serializeNode(subnode, node, index);
        index += 1;
      }.bind(this));

      this.emit(new _events.SequenceEndEvent());
  } else if ($$.isInstanceOf(node, _nodes.MappingNode)) {
      implicit = (node.tag === this.resolve(_nodes.MappingNode, node.value, true));
      this.emit(new _events.MappingStartEvent(alias, node.tag, implicit, node.flowStyle));

      $$.each(node.value, function (pairs) {
        this.serializeNode(pairs[0], node, null);
        this.serializeNode(pairs[1], node, pairs[0]);
      }.bind(this));

      this.emit(new _events.MappingEndEvent());
  }
};


module.exports.Serializer = Serializer;


////////////////////////////////////////////////////////////////////////////////
// vim:ts=2:sw=2
////////////////////////////////////////////////////////////////////////////////


lib/js-yaml/representer.js
'use strict';


var $$ = require('./core'),
    _errors = require('./errors'),
    _nodes = require('./nodes');


function RepresenterError() {
  _errors.YAMLError.apply(this, arguments);
  this.name = 'RepresenterError';
}
$$.inherits(RepresenterError, _errors.YAMLError);


function getTypeOf(obj) {
  switch (typeof obj) {
    case 'undefined': return 'null';
    case 'string':    return 'string';
    case 'boolean':   return 'boolean';
    case 'number':    return (obj === obj.toFixed()) ? 'integer' : 'float';
  }

  if (null === obj) {
    return 'null';
  }

  throw new RepresenterError('Not fully implemented yet');  
}


////////////////////////////////////////////////////////////////////////////////


function BaseRepresenter(defaultStyle, defaultFlowStyle) {
  this.defaultStyle = defaultStyle || null;
  this.defaultFlowStyle = defaultFlowStyle || null;
  this.representedObjects = {};
  this.objectKeeper = [];
  this.aliasKey = null;
}

BaseRepresenter.yamlRepresenters = {};
BaseRepresenter.addRepresenter = function addRepresenter(dataType, representer) {
  this.yamlRepresenters[dataType] = representer;
};


BaseRepresenter.prototype.represent = function represent(data) {
  var node = this.representData(data);
  this.serialize(node);
  this.representedObjects = {};
  this.objectKeeper = [];
  this.aliasKey = null;
};


BaseRepresenter.prototype.representData = function representData(data) {
  var node, dataType;

  this.aliasKey = this.ignoreAliases(data) ? null : $$.id(data);

  if (null !== this.aliasKey) {
    if (undefined !== this.representedObjects[this.aliasKey]) {
      return this.representedObjects[this.aliasKey];
    }

    this.objectKeeper.push(data);
  }

  dataType = getTypeOf(data);

  if (undefined !== this.yamlRepresenters[dataType]) {
    return this.yamlRepresenters[dataType](data);
  }

  return new _nodes.ScalarNode(data.toString());
};


////////////////////////////////////////////////////////////////////////////////


function SafeRepresenter() {
  BaseRepresenter.apply(this, arguments);
  this.yamlRepresenters = SafeRepresenter.yamlRepresenters;
}

$$.inherits(SafeRepresenter, BaseRepresenter);
SafeRepresenter.yamlRepresenters = $$.extend({}, BaseRepresenter.yamlRepresenters);
SafeRepresenter.addRepresenter = BaseRepresenter.addRepresenter;


SafeRepresenter.prototype.representNull = function representNull(data) {
  return new _nodes.ScalarNode('~');
};


SafeRepresenter.addRepresenter('null',    SafeRepresenter.prototype.representNull);
SafeRepresenter.addRepresenter('string',  SafeRepresenter.prototype.representString);
SafeRepresenter.addRepresenter('boolean', SafeRepresenter.prototype.representBoolean);
SafeRepresenter.addRepresenter('integer', SafeRepresenter.prototype.representInteger);
SafeRepresenter.addRepresenter('float',   SafeRepresenter.prototype.representFloat);
SafeRepresenter.addRepresenter('object',  SafeRepresenter.prototype.representObject);


////////////////////////////////////////////////////////////////////////////////


function Representer() {
  SafeRepresenter.apply(this, arguments);
  this.yamlRepresenters = Representer.yamlRepresenters;
}

$$.inherits(Representer, SafeRepresenter);
Representer.yamlRepresenters = $$.extend({}, SafeRepresenter.yamlRepresenters);
Representer.addRepresenter = SafeRepresenter.addRepresenter;


////////////////////////////////////////////////////////////////////////////////


module.exports.BaseRepresenter = BaseRepresenter;
module.exports.SafeRepresenter = SafeRepresenter;
module.exports.Representer = Representer;


////////////////////////////////////////////////////////////////////////////////
// vim:ts=2:sw=2
////////////////////////////////////////////////////////////////////////////////


lib/js-yaml/common.js
'use strict';


var $$ = module.exports = {};


// UTILITY METHODS
////////////////////////////////////////////////////////////////////////////////

// mimics Pythons id() method: http://docs.python.org/library/functions.html#id
// looks likt current inplementation sucks... hope this will be enought though.
//    >>> str = 'a'
//    >>> id(str)
//    3074575944L
//    >>> id('a')
//    3074575944L
//    >>> str += 'b'
//    >>> id(str)
//    3074120136L
//    >>> id('ab')
//    3074120952L
//    >>> id('ab')
//    3074120928L
//    >>> id(str)
//    3074120136L
$$.id = (function (__id__) {
  return function id(obj) {
    if (!obj.__id__) {
      obj.__id__ = __id__;
      __id__ += 1;
    }

    return obj.__id__;
  };
}(1));


// <object> $$.extend(receiver, *sources)
//
// Copy all of the properties in the source objects over to the target object.
// It's in-order, so the last source will override properties of the same name
// in previous arguments.
//
// Example: var o = $$.extend({}, a, b, c);
$$.extend = function extend(receiver) {
  var i, l, key, skip = [];

  receiver = receiver || {};
  l = arguments.length;

  if (!!arguments[l - 1] && !!arguments[l - 1].except) {
    skip = arguments[l - 1].except;
    l -= 1;
  }

  for (i = 1; i < l; i += 1) {
    if (!!arguments[i] && 'object' === typeof arguments[i]) {
      for (key in arguments[i]) {
        if (arguments[i].hasOwnProperty(key) && -1 === skip.indexOf(key)) {
          receiver[key] = arguments[i][key];
        }
      }
    }
  }

  return receiver;
};


// simple inheritance algorithm
$$.inherits = function inherits(child, parent) {
  var InheritanceGlue = function () {};

  InheritanceGlue.prototype = parent.prototype;
  child.prototype = new InheritanceGlue();

  // copy parent prototype' methods into child, so mixing made possible
  // think of it as of caching of parent's prototype methods in child
  $$.extend(child.prototype, parent.prototype, {except: [
    'arguments', 'length', 'name', 'prototype', 'caller'
  ]});

  // restore constructor
  $$.extend(child.prototype, {constructor: child});

  child.__parent__ = parent;
};


// wrapper for instanceof that allows to check inheritance after $$.inherits
$$.isInstanceOf = function isInstanceOf(obj, klass) {
  var parent;

  if (obj instanceof klass) {
    return true;
  }

  if (!!obj && !!obj.constructor) {
    parent = obj.constructor.__parent__;
    return (parent === klass || $$.isInstanceOf(parent, klass));
  }

  return false;
};


// iterates through all object keys-value pairs calling iterator on each one
// example: $$.each(objOrArr, function (val, key) { /* ... */ });
$$.each = function each(obj, iterator, context) {
  var keys, i, l;

  if (null === obj || undefined === obj) {
    return;
  }

  context = context || iterator;

  if (obj.forEach === Array.prototype.forEach) {
    obj.forEach(iterator, context);
  } else {
    keys = Object.getOwnPropertyNames(obj);
    for (i = 0, l = keys.length; i < l; i += 1) {
      iterator.call(context, obj[keys[i]], keys[i], obj);
    }
  }
};


// returns reversed copy of array
$$.reverse = function reverse(arr) {
  var result = [], i, l;
  for (i = 0, l = arr.length; i < l; i += 1) {
    result.unshift(arr[i]);
  }
  return result;
};


// Modified from:
// https://raw.github.com/kanaka/noVNC/d890e8640f20fba3215ba7be8e0ff145aeb8c17c/include/base64.js
$$.decodeBase64 = (function () {
  var padding = '=', binTable = [
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63,
    52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1,  0, -1, -1,
    -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
    15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
    -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
    41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1
  ];

  return function decode(data) {
    var value, code, idx = 0, result = [], leftbits, leftdata;

    leftbits = 0; // number of bits decoded, but yet to be appended
    leftdata = 0; // bits decoded, but yet to be appended

    // Convert one by one.
    for (idx = 0; idx < data.length; idx += 1) {
      code = data.charCodeAt(idx);
      value = binTable[code & 0x7F];

      // Skip LF(NL) || CR
      if (0x0A !== code && 0x0D !== code) {
        // Fail on illegal characters
        if (-1 === value) {
          throw new Error("Illegal characters (code=" + code + ") in position " +
                          idx + ": ordinal not in range(0..128)");
        }

        // Collect data into leftdata, update bitcount
        leftdata = (leftdata << 6) | value;
        leftbits += 6;

        // If we have 8 or more bits, append 8 bits to the result
        if (leftbits >= 8) {
          leftbits -= 8;
          // Append if not padding.
          if (padding !== data.charAt(idx)) {
            result.push((leftdata >> leftbits) & 0xFF);
          }
          leftdata &= (1 << leftbits) - 1;
        }
      }
    }

    // If there are any bits left, the base64 string was corrupted
    if (leftbits) {
      throw new Error("Corrupted base64 string");
    }

    return new Buffer(result);
  };
}());


// CLASSES
////////////////////////////////////////////////////////////////////////////////


// Dummy alternative of delayed population based on generators in PyYAML
$$.Populator = function Populator(data, callback, context) {
  if (!(this instanceof $$.Populator)) {
    return new $$.Populator(data, callback, context);
  }

  this.data = data;
  this.execute = function () {
    callback.call(context || callback);
  };
};


// Simple implementation of hashtable
$$.Hash = function Hash(defaultValue) {
  var keys, values, index;

  if (!(this instanceof $$.Hash)) {
    return new $$.Hash(defaultValue);
  }

  index = 0;
  keys = [];
  values = [];

  this.store = function store(key, value) {
    var i = keys.indexOf(key);

    if (0 <= i) {
      values[i] = value;
      return;
    }

    i = index;
    index += 1;

    keys[i] = key;
    values[i] = value;
  };


  this.remove = function remove(key) {
    var i = keys.indexOf(key);

    if (0 <= i) {
      delete keys[i];
      delete values[i];
    }
  };


  this.hasKey = function hasKey(key) {
    return 0 <= keys.indexOf(key);
  };


  this.get = function get(key) {
    var i = keys.indexOf(key);
    return (0 <= i) ? values[i] : defaultValue;
  };
};


////////////////////////////////////////////////////////////////////////////////
// vim:ts=2:sw=2
////////////////////////////////////////////////////////////////////////////////
